### DISCUSSION TITLE
Where is the part inter_feat, user_feat and item_feat are all merged together?

### DISCUSSION THREAD

**Discussion by Jeriousman** [RUCAIBox/RecBole#1213] [Q&A]
I am checking your code now to see how you are dealing with data. But I am kinda lost now as code is quite a bit lengthy. Which part of your code is the part where  inter_feat, user_feat and item_feat are all merged together? I wish you can pinpoint it for me.

Thank you in advance.

**Comment by Sherry-XLL** [comment]
@Jeriousman Hello, the data processing part of `inter_feat`, `user_feat` and `item_feat` is in `dataset.py`, but I don't quite understand what you mean by `merge together`? `inter_feat` stores the interaction features, while `user_feat` and `item_feat` store the user and item features. They are three independent data structure and are not merged together.

**Reply by Jeriousman** [reply]
In the end, in dataloader, all the features from .inter, .user, .item will all be in dataloader which means they are merged into one data variable somewhere. Can you show me the code snippet where it happens?

Thank you.

**Reply by hyp1231** [reply]
Hi,

Certainly!

* For general / context-aware / knowledge-based recommendation, the features are merged in our `Dataset`.
https://github.com/RUCAIBox/RecBole/blob/4f761699454fd5d8873607173d2514d38ac24031/recbole/data/dataset/dataset.py#L1256-L1269

* For sequential recommendation, it's a bit hard to join item features to the item sequences (for the efficiency concern). So we prefer to not merge them, but to fetch the item features via item ID in sequences in **models**. An example model that can be referred to is [DIN](https://github.com/RUCAIBox/RecBole/blob/master/recbole/model/sequential_recommender/din.py).

**Reply by Jeriousman** [reply]
Thank you for your clarification! I will check those. The library is awesome and vase and because it is vast, I am lost quite often. Your help clears the way. Thank you again.
